### PR TITLE
Optimize writable state setting judgment conditions

### PR DESCRIPTION
**Pull request by brucelwl**
Optimize writable state setting judgment conditions, and please see issue https://github.com/netty/netty/issues/9771

### CODE REVIEW

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -172,7 +172,7 @@ private void incrementPendingOutboundBytes(long size, boolean invokeLater) {
         }
 
         long newWriteBufferSize = TOTAL_PENDING_SIZE_UPDATER.addAndGet(this, size);
-        if (newWriteBufferSize > channel.config().getWriteBufferHighWaterMark()) {
+        if (newWriteBufferSize > channel.config().getWriteBufferHighWaterMark() && isWritable()) {
```
Comment (johnou) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: is isWritable a non volatile boolean check? if yes, it should go first.

Comment (johnou) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: (cheaper expressions should come first)

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 1 above)
Comment (brucelwl) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: ``newWriteBufferSize > channel.config().getWriteBufferHighWaterMark()`` is a statement that must be executed, but ``isWritable()`` is not, So I think ``newWriteBufferSize > channel.config().getWriteBufferHighWaterMark()`` should be the first

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: both are volatile so I think it not really matters.

Comment (brucelwl) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: Can this PR be merged? @normanmaurer

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: @brucelwl I want to at least have one more person to review it.

Comment (franz1981) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: I'm travelling atm and will come back on late Friday: I would like to review this so please, if there is no hurry, wait for me so I can play a bit with it :)

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: @franz1981 sounds like a plan... safe travels!

**Comment by normanmaurer** [comment]
@njhill PTAL as well

**Comment by brucelwl** [comment]
> I'm not really familiar with this code, but it looks like in theory this could result in a behaviour change w.r.t. "user defined writability".
> 
> IMHO a cleaner way to address this would be to add this right before the `compareAndSet` in `setWritable`/`setUnwritable`, WDYT?
> 
> ```java
> if (oldValue == newValue) {
>     return;
> }
> ```
I think each has its own advantages and disadvantages

**Comment by njhill** [comment]
@brucelwl fair enough :) but it still looks to me like your change is functionally incorrect as stands - the `unwritable` int field might not be updated in cases that it should be.

**Comment by brucelwl** [comment]
@njhill Can you give me more details ? Maybe your solution will have the same problem, I hope that more people can participate in the discussion of this issue and make a final decision on modification. @normanmaurer @johnou @franz1981

**Comment by njhill** [comment]
@brucelwl for example, imagine the value of `unwritable` is 4 (2nd user-defined index set) and then the outbound buffer becomes full. Before, this would cause `unwritable` to be updated to 5 (LSB set), but now that would be prevented that because `isWritable()` is already false.

Later, imagine 2nd user-defined writability index gets unset. Previously this would have meant `unwritable` changing from 5 to 1, but overall writability doesn't change so no event would be fired. With this PR, the value would change from 4 to 0, and trigger a (incorrect) writability changed event.

**Comment by brucelwl** [comment]
@njhill I just judge whether the current state to be modified is the same as the original state, and if it is the same, I will not continue

**Comment by brucelwl** [comment]
@njhill It seems that you are right, My original idea was to reduce this CAS operation

**Comment by njhill** [comment]
@brucelwl I understood your intention, the change I suggested would also do that but without any functional regression.

Here is a test case which this PR fails (can be added to `ChannelOutboundBufferTest`):

```java
    @Test
    public void testMixedWritability2() {
        final StringBuilder buf = new StringBuilder();
        EmbeddedChannel ch = new EmbeddedChannel(new ChannelInboundHandlerAdapter() {
            @Override
            public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
                buf.append(ctx.channel().isWritable());
                buf.append(' ');
            }
        });

        ch.config().setWriteBufferLowWaterMark(128);
        ch.config().setWriteBufferHighWaterMark(256);

        ChannelOutboundBuffer cob = ch.unsafe().outboundBuffer();

        // Trigger channelWritabilityChanged() by setting a user-defined writability flag to false
        cob.setUserDefinedWritability(1, false);
        ch.runPendingTasks();
        assertThat(buf.toString(), is("false "));

        // Ensure writing too many bytes does not trigger channelWritabilityChanged()
        // because of the user-defined writability flag.
        ch.write(buffer().writeZero(257));
        assertThat(buf.toString(), is("false "));

        // Ensure that setting the user-defined writability flag to true does not trigger channelWritabilityChanged()
        cob.setUserDefinedWritability(1, true);
        ch.runPendingTasks();
        assertThat(buf.toString(), is("false "));

        safeClose(ch);
    }
```

**Comment by brucelwl** [comment]
@njhill Thank you for your suggestion. I will close this pr. do you want to modify this problem yourself, or I will resubmit a pr ?

**Comment by brucelwl** [comment]
@njhill Do you think it is necessary to continue to do such optimization?

**Comment by franz1981** [comment]
@bruselwl Just saw your discussion with @njhill: will look forward the new submitted PR to review instead :)

**Comment by njhill** [comment]
@brucelwl feel free to submit another PR

**Pull request closed without merging**